### PR TITLE
net: Add option `-enablebip61` to configure sending of BIP61 notifications

### PR DESCRIPTION
**Pull request by laanwj**
This commit adds a boolean option `-peersendreject`, defaulting to `1`, that can be used to disable the sending of [BIP61](https://github.com/bitcoin/bips/blob/master/bip-0061.mediawiki) `reject` messages. This functionality has been requested for various reasons:

- security (DoS): reject messages can reveal internal state that can be used to target certain resources such as the mempool more easily.

- bandwidth: a typical node sends lots of reject messages; this counts against upstream bandwidth. Also the reject messages tend to be larger than the message that was rejected.

On the other hand, reject messages can be useful while developing client software (I found them indispensable while creating bitcoin-submittx), as well as for our own test cases, so whatever the default becomes on the long run, IMO the functionality should be retained as option. But that's a discussion for later, for now it's simply a node operator decision.

Also adds a RPC test that checks the functionality.

### CODE REVIEW

File: test/functional/p2p_invalid_tx.py [hunk 1]
```diff
@@ -136,6 +138,14 @@ def run_test(self):
         wait_until(lambda: 1 == len(node.getpeerinfo()), timeout=12)  # p2ps[1] is no longer connected
         assert_equal(expected_mempool, set(node.getrawmempool()))
 
+        # restart node with sending BIP61 messages disabled, check that it disconnects without sending the reject message
+        self.log.info('Test a transaction that is rejected, with BIP61 disabled')
+        self.restart_node(0, ['-peersendreject=0','-persistmempool=0'])
+        self.reconnect_p2p(num_connections=1)
+        txr = create_transaction(block1.vtx[0], 0, b'\x64', 50 * COIN - 12000)
+        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True)
```
[the remark this answers is not in the document]
Comment (laanwj) [reply test/functional/p2p_invalid_tx.py]: yes, it's not necessary, we can re-use tx1

**Comment by sdaftuar** [comment]
Concept ACK

**Comment by gmaxwell** [comment]
Concept ACK

**Comment by sipa** [comment]
Concept ACK

**Comment by jonasschnelli** [comment]
utACK ca4d71559c9d0f4a5a9a1dec105bff827b530a00 (after squashing the squashme)

**Comment by laanwj** [comment]
Squashed ca4d715 → 39d4d10

File: test/functional/p2p_invalid_tx.py [hunk 2]
```diff
@@ -71,7 +73,7 @@ def run_test(self):
         # and we get disconnected immediately
         self.log.info('Test a transaction that is rejected')
         tx1 = create_transaction(block1.vtx[0], 0, b'\x64', 50 * COIN - 12000)
-        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True)
+        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True, reject_code=REJECT_INVALID)
```
Comment (MarcoFalke) [inline test/functional/p2p_invalid_tx.py]: nit: If you set the reject_code, you might as well set the reject_reason:

```py
reject_reason=b'mandatory-script-verify-flag-failed (Invalid OP_IF construction)'
```

File: test/functional/p2p_invalid_tx.py [hunk 3]
```diff
@@ -136,6 +138,13 @@ def run_test(self):
         wait_until(lambda: 1 == len(node.getpeerinfo()), timeout=12)  # p2ps[1] is no longer connected
         assert_equal(expected_mempool, set(node.getrawmempool()))
 
+        # restart node with sending BIP61 messages disabled, check that it disconnects without sending the reject message
+        self.log.info('Test a transaction that is rejected, with BIP61 disabled')
+        self.restart_node(0, ['-peersendreject=0','-persistmempool=0'])
```
Comment (MarcoFalke) [inline test/functional/p2p_invalid_tx.py]: micro-style-nit: Could add a white space after the comma?

**Review by maflcko — COMMENTED** [review]
utACK 39d4d10bc961ae939b6bdabfd030140012c8b395. Just two style nits in tests

**Review by jnewbery — COMMENTED** [review]
Tested ACK 39d4d10bc961ae939b6bdabfd030140012c8b395

My only micro-nit has been covered by @MarcoFalke 

A couple of general points/questions:

- I think `-enablebip61` is a better argument name than `-peersendreject`, which I think could be a little ambiguous/confusing.
- Should this option also disable logging of received REJECT messages in `ProcessMessage()` (to shut down the possibility of peers spamming our debug log with REJECT messages).

Possible enhancements for future PRs:
- disable REJECT messages by default.
- make this configurable per peer (eg for whitelisted peers)

**Comment by MarcoFalke** [comment]
Agree that mentioning bip61 in the option name makes sense.

**Comment by laanwj** [comment]
> I think -enablebip61 is a better argument name than -peersendreject, which I think could be a little ambiguous/confusing.

As you might have guessed from the variable name I used that name at first, but thought this name was more direct/clear. But I'm ok with changing it to that.

> Should this option also disable logging of received REJECT messages in ProcessMessage() (to shut down the possibility of peers spamming our debug log with REJECT messages).

I think that's an orthogonal concern. One might want their node to be silent but still listen to other's reject messages. 
Also: The messages are not enabled unless the very noisy ::NET category (which logs multiple things for every packet!) is already enabled. So I don't think this makes log flooding a worse threat in practice. Might be something to keep in mind for #12219 though.

**Comment by MarcoFalke** [comment]
Needs rebase if still relevant

**Comment by laanwj** [comment]
Why would this already not be relevant now?

**Comment by MarcoFalke** [comment]
Sorry, this is an automated template reply :)

**Comment by laanwj** [comment]
Rebased and renamed the option.

**Comment by jnewbery** [comment]
Strange. p2p_invalid_tx.py fails in one job in Travis, on the updated line:

```
        node.p2p.send_txs_and_test([tx1], node, success=False, expect_disconnect=True, reject_code=REJECT_INVALID)
```

I can't reproduce this locally.

**Comment by laanwj** [comment]
Right - and that's not even with changed `-enablebip61` setting, but the default one
```
2018-05-09T19:56:46.996000Z TestFramework (INFO): Test a transaction that is rejected
2018-05-09T19:57:47.063000Z TestFramework.utils (ERROR): wait_until() failed. Predicate: (['            wait_until(lambda: self.reject_code_received == reject_code, lock=mininode_lock)\n'], 592)
```
If it gets to that `wait_until`, at the node will have closed the connection, but without sending a reject message.
Can't reproduce it locally either, will restart the job.

**Comment by laanwj** [comment]
I'd missed @MarcoFalke 's comment about testing reject reason, patched and squashed:
47d858ddfc481df811fb9c67a5c105c3dcbdfa10 →  56d8a5b0b58afc978bf47d7f18164d0362a433b0

**Comment by MarcoFalke** [comment]
Yes, the travis failure is unrelated. I will open an issue when I see this again.

**Comment by MarcoFalke** [comment]
re-utACK 56d8a5b0b58afc978bf47d7f18164d0362a433b0 (Only change is rebase and test fixup as well as a change to the option name)

**Comment by MarcoFalke** [comment]
Needs rebase due to merge conflict in tests

**Comment by laanwj** [comment]
Rebased 56d8a5b0b58afc978bf47d7f18164d0362a433b0 → 87fe292d897e09e176ac7e254144466c319cc9ac

**Comment by MarcoFalke** [comment]
re-utACK 87fe292 (no changes other than rebase)

**Comment by MarcoFalke** [comment]
@jonasschnelli @sdaftuar Mind to re-ACK?

**Comment by sipa** [comment]
utACK 87fe292d897e09e176ac7e254144466c319cc9ac

**Pull request merged**